c++/src/capnp/serialize.c++

```diff
@@ -272,6 +272,26 @@ void readMessageCopy(kj::InputStream& input, MessageBuilder& target,
 }
 
 // -------------------------------------------------------------------
+kj::Array<word> serializeSegmentTable(kj::ArrayPtr<const kj::ArrayPtr<const word>> segments) {
+  KJ_REQUIRE(segments.size() > 0, "Tried to serialize uninitialized message.");
+  
+  auto result = kj::heapArray<word>(segments.size() / 2 + 1);
+  auto table = reinterpret_cast<_::WireValue<uint32_t>*>(result.begin());
+
+  // We write the segment count - 1 because this makes the first word zero for single-segment
+  // messages, improving compression.  We don't bother doing this with segment sizes because
+  // one-word segments are rare anyway.
+  table[0].set(segments.size() - 1);
+  for (uint i = 0; i < segments.size(); i++) {
+    table[i + 1].set(segments[i].size());
+  }
+  if (segments.size() % 2 == 0) {
+    // Set padding byte.
+    table[segments.size() + 1].set(0);
```

Comment:
I don't think this is necessary. The `heapArray` will make all the words null right? So these 4 bytes of padding will be null already, right? If it doesn't, you can make it branchless by setting the last element in the word table 0 at the start.

Reply:
I think it depends on the behavior of the allocator?  It looks like:
- `kj::heapArray<word>()`
- calls `_::HeapArrayDisposer::allocate<word>(size)`
- calls `HeapArrayDisposer::Allocate_<word, true, true>::allocate(count, count)`
- calls `_::HeapArrayDisposer::allocateImpl(sizeof(word), elementCount, capacity, nullptr, nullptr)`
- which basically just returns `operator new(elementSize * capacity)`

I don't think `operator new()` is guaranteed to zero-initialize memory, although I wouldn't be surprised if some allocators have a mode that does, for the security/reproducibility benefits?  I notice that there's a separate `_::HeapArrayDisposer::allocateUninitialized()` that makes an almost identical call to `allocateImpl()`, except with a zero `elementCount`.

Reply:
Right, kj::heap does not guarantee memory is zero'd, and quite often it isn't.